networkit/cpp/centrality/TopHarmonicCloseness.cpp

```diff
@@ -5,535 +5,633 @@
-                       nodesPerLev[j];
-      }
+void TopHarmonicCloseness::computeReachableNodes() {
+    if (G->isDirected()) {
+        wccPtr = std::unique_ptr<WeaklyConnectedComponents>(new WeaklyConnectedComponents(*G));
+        wccPtr->run();
+        const auto compSizes = wccPtr->getComponentSizes();
+        G->parallelForNodes(
+            [&](node u) { reachableNodes[u] = compSizes.at(wccPtr->componentOfNode(u)); });
     } else {
-      for (count j = 0; j <= nLevs; j++) {
-        level_bound += inverseDistance(std::max(2., (double)j - (double)i)) *
-                       nodesPerLev[j];
-      }
+        ConnectedComponents cc(*G);
+        cc.run();
+        const auto compSizes = cc.getComponentSizes();
+        G->parallelForNodes(
+            [&](node u) { reachableNodes[u] = compSizes.at(cc.componentOfNode(u)); });
     }
+}
 
-    for (count j = 0; j < levels[i].size(); ++j) {
-      node w = levels[i][j];
-      double bound =
-          (level_bound - inverseDistance(2.) +
-           (inverseDistance(1.) - inverseDistance(2.)) * G.degree(w));
-
-      if (bound < S2[w] &&
-          (!G.isDirected() || component[w] == component[source])) {
-        S2[w] = bound;
-      }
+void TopHarmonicCloseness::run() {
+    computeReachableNodes();
+    if (useNBbound)
+        runNBbound();
+    else
+        runNBcut();
+
+    topKNodes.resize(k + trail.size());
+    topKScores.resize(k + trail.size());
+    count i = k;
+    do {
+        --i;
+        const node u = topKNodesPQ.extract_top();
+        topKNodes[i] = u;
+        topKScores[i] = hCloseness[u];
+    } while (!topKNodesPQ.empty());
+
+    for (; i < trail.size(); ++i) {
+        topKNodes[k + i] = trail[i];
+        topKScores[k + i] = hCloseness[trail[i]];
     }
-  }
-}
 
-void TopHarmonicCloseness::init() {
-  n = G.upperNodeIdBound();
-  assert(n >= k);
-
-  topk.clear();
-  topk.resize(k);
-  topkScores.clear();
-  topkScores.resize(k);
-  nMinCloseness = 0;
-  minCloseness = std::numeric_limits<double>::max();
-  trail = 0;
+    hasRun = true;
 }
 
-void TopHarmonicCloseness::run() {
-  init();
+void TopHarmonicCloseness::runNBcut() {
+    if (G->isWeighted())
+        G->parallelForNodes([&](node u) { hCloseness[u] = initialBoundNBcutWeighted(u); });
+    else
+        G->parallelForNodes([&](node u) { hCloseness[u] = initialBoundNBcutUnweighted(u); });
 
-  std::vector<bool> toAnalyze(n, true);
+    prioQ.build_heap(G->nodeRange().begin(), G->nodeRange().end());
 
-  // We compute the number of reachable nodes (or an upper bound)
-  // only if we use the algorithm for complex networks.
-  if (!useBFSbound) {
-    computeReachableNodes();
-  }
+    std::atomic_bool stop{false};
+    std::atomic<double> kthCloseness{-1};
 
-  // Main priority queue with all nodes in order of decreasing degree
-  Aux::PrioQueue<edgeweight, node> Q1(n);
+#pragma omp parallel
+    while (!stop.load(std::memory_order_relaxed)) {
```

Comment:
Here I am trying to avoid race conditions on `stop` while trying to avoid synchronization overheads. `stop` can only be set to `true` (i.e., when the algorithm can finish) and not vice versa, and there is no harm (apart from some loss in performance) if more iterations are done.

Reply:
Maybe I am wrong here, but with memory_order_relaxed there can be race conditions.

https://en.cppreference.com/w/cpp/atomic/memory_order#Explanation
https://stackoverflow.com/questions/48139399/memory-order-relaxed-not-work-as-expected-in-code-from-c-concurrency-in-action

Reply:
There can be race conditions (in the sense that the code doesn't "see" the current version of other variables such as `kthCloseness`; this is *not* a "data race" in C++'s terms) but that's not an issue if the `stop` variable is only used for stopping. When the `#pragma omp parallel` scope is exited, there is another (implicit) memory barrier that makes sure that we see all

A plain Boolean would not be conforming to the C++ standard (but likely still work in practice here): if we use a plain Boolean, there can be a *data race* (i.e., unsychronized concurrent access) but is UB according to C++.

Reply:
Good to know, thx!